libfoundation/src/system-library-android.hpp

```diff
@@ -0,0 +1,168 @@
+/*                                                                     -*-c++-*-
+ Copyright (C) 2017 LiveCode Ltd.
+ 
+ This file is part of LiveCode.
+ 
+ LiveCode is free software; you can redistribute it and/or modify it under
+ the terms of the GNU General Public License v3 as published by the Free
+ Software Foundation.
+ 
+ LiveCode is distributed in the hope that it will be useful, but WITHOUT ANY
+ WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ for more details.
+ 
+ You should have received a copy of the GNU General Public License
+ along with LiveCode.  If not see <http://www.gnu.org/licenses/>.  */
+
+#include "system-private.h"
+
+#include <foundation-auto.h>
+
+/* ================================================================
+ * Android Handle Class
+ * ================================================================ */
+
+#include "system-library-posix.hpp"
+
+#include <dlfcn.h>
+#include <link.h>
+
+#if __ANDROID_API__ < 21
+extern "C" __attribute__((weak)) int
+dl_iterate_phdr(int (*)(struct dl_phdr_info *, size_t, void *),
+                void *);
+#endif
+
+extern MCStringRef gMCSAndroidNativeLibPath;
+
+class __MCSLibraryHandleAndroid: public __MCSLibraryHandlePosix
+{
+public:
+    bool
+    CopyNativePath(MCStringRef& r_native_path) const
+    {
+        const char *t_sys_filename = nullptr;
+        if (dl_iterate_phdr == nullptr)
+        {
+            t_sys_filename = CopyNativePathLegacy();
+        }
+        else
+        {
+            t_sys_filename = CopyNativePathModern(m_handle).Run();
+        }
+
+        if (t_sys_filename != nullptr)
+        {
+            
+            MCAutoStringRef t_filename;
+            if (!MCStringCreateWithSysString(t_sys_filename,
+                                             &t_filename))
+            {
+                return false;
+            }
+        
+            return MCStringFormat(r_native_path,
+                                  "%@/%@",
+                                  gMCSAndroidNativeLibPath,
+                                  *t_filename);
+        }
+
+        return __MCSLibraryThrowResolvePathFailed();
+    }
+    
+private:
+    class CopyNativePathModern
+    {
+    public:
+        CopyNativePathModern(void *p_handle)
+            : m_handle(p_handle),
+              m_filename(nullptr)
+        {
+        }
+        
+        const char *Run(void)
+        {
+            dl_iterate_phdr(Thunk,
+                            this);
+            return m_filename;
+        }
+        
+    private:
+        int operator () (struct dl_phdr_info* p_info,
+                         size_t p_size)
+        {
+            if (m_filename != nullptr)
+            {
+                return 0;
+            }
+            
+            Dl_info t_addr_info;
+            if (dladdr(reinterpret_cast<void *>(p_info->dlpi_addr),
```

Comment:
As I understand it, this is a [`static_cast`](http://en.cppreference.com/w/cpp/language/static_cast) because:

> 10) A prvalue of type pointer to `void` (possibly cv-qualified) can be converted to pointer to any type. If the value of the original pointer satisfies the alignment requirement of the target type, then the resulting pointer value is unchanged, otherwise it is unspecified. Conversion of any pointer to pointer to `void` and back to pointer to the original (or more cv-qualified) type preserves its original value.